### PR TITLE
Add namespace and assembly check before interception

### LINKED ISSUE #51233: RequestDelegateGenerator intercepts incorrect methods
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

When `RequestDelegateGenerator` decides whether it should intercept an endpoint call it uses the name of the method (_KnownMethods_) and part of the signature (Last argument should be a `System.Delegate`). This makes it intercept method calls that uses the same signature but that are not ment to be intercepted.

Easiest fix would be to just check the containing assembly inside the [`InvocationOperationExtensions.IsValidOperation`](https://github.com/dotnet/aspnetcore/blob/main/src/Http/Http.Extensions/gen/StaticRouteHandlerModel/InvocationOperationExtensions.cs#L26) method or something similar

E.g.

```C#
private const string KnownAssembly = "Microsoft.AspNetCore.Routing";

public static readonly string[] KnownMethods =
{
...
};

public static bool IsValidOperation(this IOperation? operation, WellKnownTypes wellKnownTypes, [NotNullWhen(true)] out IInvocationOperation? invocationOperation)
{
    invocationOperation = null;
    if (operation is IInvocationOperation targetOperation &&
        targetOperation.Type?.ContainingAssembly.Identity.Name == KnownAssembly &&
....
```

### Expected Behavior

Only _Map_ methods defined inside the `Microsoft.AspNetCore.Builder` namespace should be intercepted. Don't only rely on the name / signature but also include the assembly/namespace or something else that distinctly tells you it's one of the methods within the aspnet core framework.

### Steps To Reproduce

Run the following sample program with Aot enabled

```C#
var builder = WebApplication.CreateSlimBuilder(args);
builder.Services.Map(() => Task.CompletedTask);
var app = builder.Build();
app.Map("route", () => Results.Text("Working"));
app.Run();

static class Ext
{
    public static IServiceCollection Map(this IServiceCollection services, Delegate @delegate) => services;
}
```

### Exceptions (if any)

```Cannot intercept method 'Ext.Map(IServiceCollection, Delegate)' with interceptor 'GeneratedRouteBuilderExtensionsCore.Map0(IEndpointRouteBuilder, string, Delegate)' because the signatures do not match.```

### .NET Version

8.0.100-rc.1.23455.8

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by captainsafia**
Closes https://github.com/dotnet/aspnetcore/issues/51233

### CODE REVIEW

File: src/Http/Http.Extensions/test/RequestDelegateGenerator/CompileTimeCreationTests.cs [hunk 1]
```diff
@@ -556,4 +556,61 @@ public async Task SupportsHandlersWithSameSignatureButDifferentParameterNamesFro
         Assert.Equal(new EventId(5, "ImplicitBodyNotProvided"), log2.EventId);
         Assert.Equal(@"Implicit body inferred for parameter ""todo1"" but no body was provided. Did you mean to use a Service instead?", log2.Message);
     }
+
+    [Fact]
+    public async Task SkipsMapGetWithIncorrectNamespaceAndAssembly()
```
Comment (BrennanConroy) [inline src/Http/Http.Extensions/test/RequestDelegateGenerator/CompileTimeCreationTests.cs:561]: super nit: MapGet -> Map
There is no MapGet being tested here

File: src/Http/Http.Extensions/test/RequestDelegateGenerator/CompileTimeCreationTests.cs [hunk 2]
```diff
@@ -556,4 +556,61 @@ public async Task SupportsHandlersWithSameSignatureButDifferentParameterNamesFro
...
+    {
+        app.ServiceProvider.Map(1, (string test) => "Hello world!");
+        app.ServiceProvider.MapPost(2, (string test) => "Hello world!");
+        return app;
+    }
+}
+
+public static class EndpointRouteBuilderExtensions
+{
+    public static IServiceProvider Map(this IServiceProvider app, int id, Delegate requestDelegate)
```
Comment (BrennanConroy) [inline src/Http/Http.Extensions/test/RequestDelegateGenerator/CompileTimeCreationTests.cs:580]: Another nit: Should use the exact same signature as Map
```
public static IEndpointConventionBuilder Map(
        this IEndpointRouteBuilder endpoints,
        RoutePattern pattern,
        RequestDelegate requestDelegate)
```

**Review by BrennanConroy — APPROVED** [review]

File: src/Http/Http.Extensions/test/RequestDelegateGenerator/CompileTimeCreationTests.cs (hunk 2 above)
Comment (captainsafia) [reply src/Http/Http.Extensions/test/RequestDelegateGenerator/CompileTimeCreationTests.cs]: Based on the overload resolution rules for extension methods, the compiler will bind to the methods defined in the framework if they have the same signature. I did update the tests to include a scenario with extension methods targeting `IEndpointRouteBuilder` and a different type for the second parameter.

**Pull request merged**